vue/custom-event-name-casing doesn't flag incorrectly cased emits in the template

<!--
  ❗Please don't ignore this template.❗

  If you ignore it, we're just going to respond asking you to fill it out, which wastes everyone's time.
  The more relevant information you can include, the faster we can find the issue and fix it without asking you for more info.
-->

<!--
  Before posting the issue, please confirm that the problem you're getting
  is not related with your code editor configuration.
  To make sure it's not, run: yarn eslint src/your-file.vue
-->

**Checklist**

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have read the [FAQ](https://eslint.vuejs.org/user-guide/#faq) and my problem is not listed.
<!-- If you do not read the FAQ and open an issue that is listed in the FAQ, we may silently close the issue. -->

**Tell us about your environment**

- **ESLint version:** 9.12.0
- **eslint-plugin-vue version:** 9.29.0
- **Vue version:** 3.5.9
- **Node version:** 18.20.4
- **Operating System:** Arch Linux

**Please show your full configuration:**
<!-- Paste content of your .eslintrc file -->
```json5
import pluginVue from 'eslint-plugin-vue'
import vueTsEslintConfig from '@vue/eslint-config-typescript'
import pluginVitest from '@vitest/eslint-plugin'
import skipFormatting from '@vue/eslint-config-prettier/skip-formatting'

export default [
  {
    name: 'app/files-to-lint',
    files: ['**/*.{ts,mts,tsx,vue}'],
  },

  {
    name: 'app/files-to-ignore',
    ignores: ['**/dist/**', '**/dist-ssr/**', '**/coverage/**'],
  },

  ...pluginVue.configs['flat/essential'],
  ...vueTsEslintConfig(),
  
  {
    ...pluginVitest.configs.recommended,
    files: ['src/**/__tests__/*'],
  },
  skipFormatting,
  {
    rules: {
      "vue/custom-event-name-casing": "error"
    }
  }
]
```

**What did you do?**	
<!-- Please include the actual source code causing the issue. -->
```vue
<script setup lang="ts">
const emit = defineEmits({ 'foo-bar': () => true })

emit('foo-bar')
</script>

<template>
  <button @click="emit('foo-bar')">Foo</button>
  <button @click="$emit('foo-bar')">Foo</button>
</template>
```

**What did you expect to happen?**
ESLint should flag all three emits because they all use kebab case instead of camelcase.

**What actually happened?**
<!--
  Please include the actual, raw output from ESLint.
  If you are only looking at the results of your editor extension, also check the CLI results.
-->
It flags the one in script setup and the one using `$emit` in the template, but not the one using `emit`.
```
➜  vue-project npm run lint

> vue-project@0.0.0 lint
> eslint . --fix


/home/.../dev/vue-project/src/components/HelloWorld.vue
  4:6   error  Custom event name 'foo-bar' must be camelCase  vue/custom-event-name-casing
  9:25  error  Custom event name 'foo-bar' must be camelCase  vue/custom-event-name-casing

✖ 2 problems (2 errors, 0 warnings)
```
**Repository to reproduce this issue**
<!--
  ❗ Please share a repository that can reproduce your issue.
  ❗ If you don't share it, we just label it as requesting a repository that can reproduce your issue.
  ❗ If you don't share it for a few days, we silently close this issue.
-->
https://github.com/henribru/eslint-plugin-vue-custom-event-name-casing-bug

--- a/lib/rules/custom-event-name-casing.js
+++ b/lib/rules/custom-event-name-casing.js
@@ -101,6 +101,7 @@ module.exports = {
   create(context) {
     /** @type {Map<ObjectExpression|Program, {contextReferenceIds:Set<Identifier>,emitReferenceIds:Set<Identifier>}>} */
     const setupContexts = new Map()
+    let emitParamName = ''
     const options =
       context.options.length === 1 && typeof context.options[0] !== 'string'
         ? // For backward compatibility
@@ -189,7 +190,11 @@ module.exports = {
             // cannot check
             return
           }
-          if (callee.type === 'Identifier' && callee.name === '$emit') {
+
+          if (
+            callee.type === 'Identifier' &&
+            (callee.name === '$emit' || callee.name === emitParamName)
+          ) {
             verify(nameWithLoc)
           }
         }
@@ -209,6 +214,8 @@ module.exports = {
             if (emitParam.type !== 'Identifier') {
               return
             }
+            emitParamName = emitParam.name
+
             // const emit = defineEmits()
             const variable = findVariable(
               utils.getScope(context, emitParam),

--- a/tests/lib/rules/custom-event-name-casing.js
+++ b/tests/lib/rules/custom-event-name-casing.js
@@ -329,6 +329,22 @@ tester.run('custom-event-name-casing', rule, {
       </script>
       `,
       options: ['kebab-case']
+    },
+    // setup defineEmits
+    {
+      filename: 'test.vue',
+      code: `
+      <script setup>
+      const emit = defineEmits({})
+      emit('foo-bar')
+      </script>
+
+      <template>
+      <button @click="emit('foo-bar')">Foo</button>
+      <button @click="$emit('foo-bar')">Foo</button>
+      </template>
+      `,
+      options: ['kebab-case']
     }
   ],
   invalid: [
@@ -603,6 +619,35 @@ tester.run('custom-event-name-casing', rule, {
         {
           message: "Custom event name 'foo-bar' must be camelCase.",
           line: 4
+        }
+      ]
+    },
+    {
+      // https://github.com/vuejs/eslint-plugin-vue/issues/2577
+      filename: 'test.vue',
+      code: `
+      <script setup>
+      const emit = defineEmits({})
+      emit('foo-bar')
+      </script>
+
+      <template>
+      <button @click="emit('foo-bar')">Foo</button>
+      <button @click="$emit('foo-bar')">Foo</button>
+      </template>
+      `,
+      errors: [
+        {
+          message: "Custom event name 'foo-bar' must be camelCase.",
+          line: 4
+        },
+        {
+          message: "Custom event name 'foo-bar' must be camelCase.",
+          line: 8
+        },
+        {
+          message: "Custom event name 'foo-bar' must be camelCase.",
+          line: 9
         }
       ]
     }

